runtime/compiler/env/DependencyTable.cpp

```diff
@@ -32,9 +32,126 @@ TR_AOTDependencyTable::TR_AOTDependencyTable(TR_J9SharedCache *sharedCache) :
+
+      auto m_it = _methodMap.insert({method, {0, methodDependencies}});
+      auto methodEntry = &(*m_it.first);
+
+      for (size_t i = 1; i <= totalDependencies; ++i)
+         {
+         bool needsInitialization = false;
+         uintptr_t chainOffset = decodeDependencyOffset(methodDependencies[i], needsInitialization);
+         uintptr_t offset = _sharedCache->startingROMClassOffsetOfClassChain(_sharedCache->pointerFromOffsetInSharedCache(chainOffset));
+         auto entry = getOffsetEntry(offset, true);
+         if (needsInitialization)
+            entry->_waitingInitMethods.insert(methodEntry);
+         else
+            entry->_waitingLoadMethods.insert(methodEntry);
+
+         if (findCandidateForDependency(entry->_loadedClasses, needsInitialization))
+            numberRemainingDependencies -= 1;
+         }
+
+      if (numberRemainingDependencies == 0)
+         {
+         stopTracking(methodEntry);
+         dependenciesSatisfied = true;
+         }
+      else
+         {
+         methodEntry->second._remainingDependencies = numberRemainingDependencies;
+         }
+      }
+   catch (std::exception&)
+      {
+      deactivateTable();
+      return false;
+      }
+
+   return true;
+   }
+
+void
+TR_AOTDependencyTable::methodWillBeCompiled(J9Method *method)
+   {
+   OMR::CriticalSection cs(_tableMonitor);
+   if (!isActive())
+      return;
+
+   // TODO: For now we simply stop tracking method if for some reason a
+   // compilation is triggered for it, but if the compilation in question is an
+   // AOT load we might consider preventing the load from taking place (by
+   // increasing the counts and continuing to track the method, or marking the
+   // method as ineligible for loads and giving up on tracking it).
+   stopTracking(method);
+   }
+
+void
+TR_AOTDependencyTable::stopTracking(MethodEntryRef entry)
+   {
+   auto methodEntry = entry->second;
+   auto dependencyChain = methodEntry._dependencyChain;
+   auto dependencyChainLength = *dependencyChain;
+
+   for (size_t i = 1; i <= dependencyChainLength; ++i)
+      {
+      bool needsInitialization = false;
+      uintptr_t chainOffset = decodeDependencyOffset(dependencyChain[i], needsInitialization);
+      uintptr_t offset = _sharedCache->startingROMClassOffsetOfClassChain(_sharedCache->pointerFromOffsetInSharedCache(chainOffset));
+
+      auto o_it = _offsetMap.find(offset);
+
+      if (needsInitialization)
+         o_it->second._waitingInitMethods.erase(entry);
+      else
+         o_it->second._waitingLoadMethods.erase(entry);
+
+      eraseOffsetEntryIfEmpty(o_it->second, offset);
+      }
+
+   _methodMap.erase(entry->first);
+   }
+
+void
+TR_AOTDependencyTable::stopTracking(J9Method *method)
+   {
+   auto entry = _methodMap.find(method);
+   if (entry != _methodMap.end())
+      stopTracking(&*entry);
+   }
+
+void
+TR_AOTDependencyTable::eraseOffsetEntryIfEmpty(const OffsetEntry &entry, uintptr_t offset)
```

Comment:
Ideally, we would have an iterator as parameter that would capture both the entry and the offset because they come same hashtable entry. Maybe refine later.